src/_pytest/stepwise.py

```diff
@@ -47,43 +64,105 @@ def pytest_sessionfinish(session: Session) -> None:
-        self.lastfailed: str | None = self.cache.get(STEPWISE_CACHE_DIR, None)
         self.skip: bool = config.getoption("stepwise_skip")
+        self.reset: bool = config.getoption("stepwise_reset")
+        self.cached_info = self._load_cached_info()
+
+    def _load_cached_info(self) -> StepwiseCacheInfo:
+        cached_dict: dict[str, Any] | None = self.cache.get(STEPWISE_CACHE_DIR, None)
+        if cached_dict:
+            try:
+                return StepwiseCacheInfo(
+                    cached_dict["last_failed"],
+                    cached_dict["last_test_count"],
+                    cached_dict["last_cache_date_str"],
+                )
+            except (KeyError, TypeError) as e:
+                error = f"{type(e).__name__}: {e}"
+                self.report_status.append(f"error reading cache, discarding ({error})")
+
+        # Cache not found or error during load, return a new cache.
+        return StepwiseCacheInfo.empty()
 
     def pytest_sessionstart(self, session: Session) -> None:
         self.session = session
 
     def pytest_collection_modifyitems(
         self, config: Config, items: list[nodes.Item]
     ) -> None:
-        if not self.lastfailed:
-            self.report_status = "no previously failed tests, not skipping."
+        last_test_count = self.cached_info.last_test_count
+        self.cached_info.last_test_count = len(items)
+
+        if self.reset:
+            self.report_status.append("resetting state, not skipping.")
+            self.cached_info.last_failed = None
+            return
+
+        if not self.cached_info.last_failed:
+            self.report_status.append("no previously failed tests, not skipping.")
+            return
+
+        if last_test_count is not None and last_test_count != len(items):
+            self.report_status.append(
+                f"test count changed, not skipping (now {len(items)} tests, previously {last_test_count})."
+            )
+            self.cached_info.last_failed = None
             return
 
-        # check all item nodes until we find a match on last failed
+        # Check all item nodes until we find a match on last failed.
         failed_index = None
         for index, item in enumerate(items):
-            if item.nodeid == self.lastfailed:
+            if item.nodeid == self.cached_info.last_failed:
                 failed_index = index
                 break
 
         # If the previously failed test was not found among the test items,
         # do not skip any tests.
         if failed_index is None:
-            self.report_status = "previously failed test not found, not skipping."
+            self.report_status.append("previously failed test not found, not skipping.")
         else:
-            self.report_status = f"skipping {failed_index} already passed items."
+            self.report_status.append(
+                f"skipping {failed_index} already passed items (cache from {self.cached_info.last_cache_date},"
```

Comment:
```suggestion
                f"skipping {failed_index} already passed items (cache from {self.cached_info.last_cache_date:%c},"
```

maybe, where `%c` is ["Locale’s appropriate date and time representation."](https://docs.python.org/3/library/datetime.html#strftime-and-strptime-format-codes)?

Without this, it gets displayed as e.g. `cache from 2025-01-15 14:25:45.458813` which seems a bit verbose, and the fractional seconds don't really add any useful information.

Reply:
I did try that, but to me it showed up as the default. IIUC we need to call `locale.setlocale` at some point, otherwise it will always use the `C` locale.

But I might be wrong, can you test that in your machine?

If `%c` does not work, we might explicitly use `%YY-%M-%d %h:%m:%s` to get rid of the fractional seconds, which I agree is not useful.

Reply:
Oh wow, looks like it! My locale setup is a bit weird at the moment anyways (experimented with [en_DK](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=137870) to get [proper dates](https://xkcd.com/1179/), but turns out it's not supported everywhere so now I get a mess of formats...). Because of that, I expected that to be an issue on my end.

What do you think about making a `timedelta` to now and then just formatting that in instead? That would then show up as `0:00:10 ago` or `42 days, 1:23:45 ago` which seems a much nicer format for the information we actually want to convey here.

Reply:
Good idea about `timedelta`, will do.

Reply:
Done.

I considered only showing that information if the cache was "old", but then what could be considered an "old" cache? 1 hr? 1 day? In the ended I kept it simple and always show the cache age for now.

Reply:
Hehe, I have been thinking about that as well, and came to the same conclusion as you did :sweat_smile: